pkg/gpu/cuda/env.go

```diff
@@ -0,0 +1,117 @@
+)
+
+const cudaVisibleDevicesEnvVar = "CUDA_VISIBLE_DEVICES"
+
+// GetVisibleDevicesForProcess modifies the list of GPU devices according to the
+// value of the CUDA_VISIBLE_DEVICES environment variable for the specified
+// process. Reference:
+// https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
+// Invalid device indexes are ignored, and anything that comes after that is
+// invisible, following the spec: "If one of the indices is invalid, only the
+// devices whose index precedes the invalid index are visible to CUDA
+// applications." If an invalid index is found, an error is returned together
+// with the list of valid devices found up until that point.
+func GetVisibleDevicesForProcess(systemDevices []nvml.Device, pid int, procfs string) ([]nvml.Device, error) {
+	cudaVisibleDevices, err := kernel.GetProcessEnvVariable(pid, cudaVisibleDevicesEnvVar, procfs)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get env var %s for process %d: %w", cudaVisibleDevicesEnvVar, pid, err)
+	}
+
+	return getVisibleDevices(systemDevices, cudaVisibleDevices)
+}
+
+// getVisibleDevices processes the list of GPU devices according to the value of
+// the CUDA_VISIBLE_DEVICES environment variable
+func getVisibleDevices(systemDevices []nvml.Device, cudaVisibleDevices string) ([]nvml.Device, error) {
+	if cudaVisibleDevices == "" {
+		return systemDevices, nil
+	}
+
+	var filteredDevices []nvml.Device
+	visibleDevicesList := strings.Split(cudaVisibleDevices, ",")
+
+	for _, visibleDevice := range visibleDevicesList {
+		var matchingDevice nvml.Device
+		var err error
+		if strings.HasPrefix(visibleDevice, "GPU-") {
+			matchingDevice, err = getDeviceWithMatchingUUIDPrefix(systemDevices, visibleDevice)
+			if err != nil {
+				return filteredDevices, err
+			}
+		} else if strings.HasPrefix(visibleDevice, "MIG-GPU") {
+			// MIG (Multi Instance GPUs) devices require extra parsing and data
+			// about the MIG instance assignment, which is not supported yet.
+			return filteredDevices, fmt.Errorf("MIG devices are not supported")
+		} else {
+			matchingDevice, err = getDeviceWithIndex(systemDevices, visibleDevice)
+			if err != nil {
+				return filteredDevices, err
+			}
+		}
+
+		filteredDevices = append(filteredDevices, matchingDevice)
+	}
+
+	return filteredDevices, nil
+}
+
+// getDeviceWithMatchingUUIDPrefix returns the first device with a UUID that
+// matches the given prefix. If there are multiple devices with the same prefix
+// or the device is not found, an error is returned.
+func getDeviceWithMatchingUUIDPrefix(systemDevices []nvml.Device, uuidPrefix string) (nvml.Device, error) {
+	var matchingDevice nvml.Device
+	var matchingDeviceUUID string
+
+	for _, device := range systemDevices {
+		uuid, ret := device.GetUUID()
+		if ret != nvml.SUCCESS {
+			return nil, fmt.Errorf("cannot get UUID for device: %s", nvml.ErrorString(ret))
+		}
+
+		if strings.HasPrefix(uuid, uuidPrefix) {
```

Comment:
The GPU device UUID is the whole thing, with the `GPU-` prefix included. It's a bit confusing because they call it UUID instead of "Prefix + UUID" because it's actually what it is and I've just replicated that naming here.